Fix #4503: Fixed RTE element view in hint editor

aks681 opened the pull request:
I have created a preview for the hint, if it is LaTex, and gave the name of the RTE element if its any other element in the short summary space for hints.

**Checklist**
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes.
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

aks681 committed:
fix #4503

aks681 committed:
Added other RTE elements

aks681 committed:
Added RTE name

kevinlee12 commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateHints.js:
```diff
@@ -31,15 +31,26 @@ oppia.controller('StateHints', [
+      for (var i = 0; i < stateHintsService.displayed.length; i++) {
+        if (
+          stateHintsService.displayed[i].hintContent.getHtml().indexOf(
+            'latex') != -1) {
```
Use `!==`, it's faster in terms of comparison.

kevinlee12 reviewed (commented):
Hmm, your code doesn't consider images, etc.

aks681 commented:
Yeah, to have all hint summary fields to be of equal size, I didn't add preview of other RTE components like images (though I did give a label as to what RTE element, the hint consists of). Should I add a preview of every RTE element then, even in the preview field?

aks681 committed:
made review change

kevinlee12 commented:
Hmm, I was trying to see how images would look, but it seems that I'm having difficulty saving hints. The order of my hints was: 
- Hint 1: Fraction (modal froze, but saved after I clicked on Cancel)
- Hint 2: Image (can't save this one)

seanlip commented:
Adding a special-case for latex is somewhat icky, actually. For now, can we just do the name of the RTE component for all RTE components? Less informative, but it's better to avoid spreading info about one particular RTE component throughout the code.

seanlip reviewed (changes requested):
(See top-level comment.)

aks681 commented:
@seanlip, Ok will change that.
@kevinlee12, I am not able to reproduce it, hints are saving correctly in mine, could it be interaction dependent?

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateHints.js:
```diff
@@ -57,6 +57,11 @@ oppia.controller('StateHints', [
         hint.hintContent.getHtml() ?
           $filter('convertToPlainText')(hint.hintContent.getHtml()) : '');
+      if (hintAsPlainText === '') {
+        hintAsPlainText = hint.hintContent.getHtml().split(
```
This seems ... wrong? What if the hint has more than one rich-text component, plus some other text, etc.?

I suggest taking a look at my comments on #4453, including the reference to [this approach](https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/expressions/ExpressionInterpolationService.js#L44) (using string.replace()). Perhaps the common functionality could go in a separate filter so that @Nix07 can use it too.

aks681 replied:
Done

kevinlee12 commented:
Yup, this is causing issues on the UI side as well.

seanlip commented:
Hm. @aks681 did you test? At this stage it's really important to do so, we don't have time left for testing again.

aks681 commented:
@seanlip, will take care of the case with multiple RTE components. Also, about testing shall I add a new filter and test it in filtersSpec.js?

seanlip commented:
Yep, sgtm. Thanks!

aks681 commented:
@seanlip , if both RTE components and standard text is there in the html, what should be shown in the hint preview? Right now, I am displaying 'Hint consists of [component1,component2]' in that field. Shall I just 'and' and display the text next to that?

aks681 committed:
made review changes

aks681 committed:
removed additional function

aks681 commented:
Currently, I am formatting the hint preview such that '[rteComponentName]' is shown at the positions where the RTE component exists in the actual hint. Added an example of the output in the comments in filter.js and added tests. @seanlip, is this fine?
Also, @kevinlee12, can you check if the issues still persist?

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -404,3 +404,34 @@ oppia.filter('stripFormatting', [function() {
+    html = html.replace(/&quot;/ig, '');
+    //Replace all html tags other than <oppia-noninteractive-**> ones to ''
+    html = html.replace(/<(?!oppia-noninteractive\s*?)[^>]+>/g, '');
+    html = html.split(/[<>]+/);
```
This varname should now be htmlArray, right? (I was looking at the subsequent code and thinking that html[i] was a character...)

aks681 replied:
Done, removed the use of htmlArray.

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -404,3 +404,34 @@ oppia.filter('stripFormatting', [function() {
+    html = html.replace(/<(?!oppia-noninteractive\s*?)[^>]+>/g, '');
+    html = html.split(/[<>]+/);
+    var formattedOutput = '';
+    for (var i = 0; i < html.length; i++) {
```
Instead of a manual loop, can you use html.replace() similar to https://github.com/oppia/oppia/blob/develop/core/templates/dev/head/expressions/ExpressionInterpolationService.js#L45 ? Or is that trickier?

seanlip replied:
(To be clear, I was thinking that the regex to look for when replacing is something that matches `<oppia-noninteractive-* ...></oppia-noninteractive-*>`.)

aks681 replied:
That may be trickier as all I am using is the text just after <oppia-noninteractive- but without the attributes following that, can a disjointed string like that be replaced using string.replace()?

seanlip replied:
I think so? Should be possible to represent it with a regex and a capture group, where the "..." stuff is basically "any character that's not '>'"

Something like `<oppia-noninteractive-([a-z]+) [^>]*></oppia-noninteractive-[a-z]+>` maybe?

aks681 replied:
Done, used a function in the replace string part of string.replace() to replace the entire tag with the name of the rte component

seanlip reviewed (commented):
The behaviour looks good! I had a couple of code comments.

aks681 committed:
made review changes

aks681 committed:
minor change

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -404,3 +404,33 @@ oppia.filter('stripFormatting', [function() {
+    html = html.replace(/&quot;/ig, '');
+    //Replace all html tags other than <oppia-noninteractive-**> ones to ''
+    html = html.replace(/<(?!oppia-noninteractive\s*?)[^>]+>/g, '');
+    formattedOutput = html.replace(/(<([^>]+)>)/g, function(rteTag) {
```
Missing var

seanlip commented on core/templates/dev/head/pages/exploration_editor/editor_tab/StateHints.js:
```diff
@@ -54,9 +54,8 @@ oppia.controller('StateHints', [
-      var hintAsPlainText = (
-        hint.hintContent.getHtml() ?
-          $filter('convertToPlainText')(hint.hintContent.getHtml()) : '');
+      var hintAsPlainText;
```
var hintAsPlainText = $filter(...)

seanlip commented on core/templates/dev/head/filters.js:
```diff
@@ -404,3 +404,33 @@ oppia.filter('stripFormatting', [function() {
+      }
+      return ' [' + replaceString + '] ';
+    });
+    if (formattedOutput[0] === ' ') {
```
Perhaps replace all this with `return formattedOutput.trim()`?

seanlip reviewed (commented):
Much better, a few more things!

seanlip commented:
We can take care of the remaining comments in a separate pass, though. I'm merging this in for the release. Thanks @aks681!

aks681 commented:
Ok, I'll create another PR for the rest.
Thanks!

aks681 committed:
made review changes
